enhancements/insights/container-scanner.md

```diff
@@ -0,0 +1,498 @@
+3. __Memory and CPU usage should not be input derived.__ In addition to achieving system stability and performance goals, ensuring the overhead of a scanning process does not increase significantly relative to input data is critical in preventing potential Denial-of-Service attacks relying on a specially crafted container filesystem.  
+
+##### No Impact on Workloads
+
+Scanning containers must not disrupt the containers that are scanned and have an impact on their quality of service.
+The container scanner is running with its own memory and CPU requirements and does not consume resources from the scanned container.
+
+##### Measurable Impact on Gathering Time & Payload
+
+The workload data is gathered every 12 hours by the Insights Operator.
+Gathering the runtime data will increase the overall executing time of the workload gatherer. The increase must remain reasonable.
+
+In the existing initial implementation of the container scanner, it takes __~180 ms to scan a single container__ (including the HTTP roundtrip).
+
+The current workflow for scanning the containers could be further optimized by performing the scanning after the container shapes have been tested for duplicates and scan only once instead of scanning every duplicated container.
+
+It could also be possible to scan containers in parallel instead of doing it iteratively.
+
+The additional data in the `workloadRuntimeInfoContainer` will increase the size of the insights payload that is archived and sent to Red Hat backends.
+
+At most, the additional data adds __~200 uncompressed bytes per WorkloadContainerShape in the JSON payload__.
+The current maximum number of pods that are checked by the Insights Operator is 8000 pods.
+
+This additional payload does not scale with the number of pods but with the number of `WorkloadContainerShape` which can be approximated as the number of applications (deployments,statefulsets,...) times the number of containers per pod.
+
+#### UX Risk & Mitigation
+
+The additional runtime info added to the gathered data will be exposed to the user though the archives generated by the Insights operator. As the values of the field are hashed, the user will not be able to determine their corresponding content.
+
+Exposing these data to the user with a clean user interface (eg within console.redhat.com) is out of scope of this proposal.
+
+### Drawbacks
+
+If implemented, Red Hat will document the addition of this new collection. We  introduce the risk that some number of customers who would have or are sending telemetry data via the Insights operator will choose to turn it off because they are uncomfortable sharing this additional data. 
+
+The container scanner needs to be continually updated to detect new workloads  that Red Hat wants information about. 
+This can be done iteratively and will require an updated reference to the container scanner image in the Insights Operator to take into account the additional detections.
```

Comment:
My feeling is that this is pretty fragile. I mean, what are the guarantees for the future that the scanner doesn't scan too many things? I guess there should be some "safe & revision" mechanism.

Reply:
The container scanner will scan for a limited set of information (`runtime`, `runtime-kind`) that are mutually exclusive. So adding more detection "values" would not increase the scanning performance. When we detect that a container is running Python, we will only run 1 fingerprint for it (and not run any fingerprint related to Java, Golang, etc.)

Reply:
Just to extend on this question: does this mean that we can only detect/measure runtimes we know? Is there no way to gather information about runtimes we don't know already?

Reply:
that's a tricky question. How would we identify a runtime we don't know with the privacy constraints we have?
In Java land, we could for example report the main class + class path but as this would be hashed and truncated before it is collected in the insights archive, it would be impossible to find the original values.
It'd be possible to report the process name but this might provide not enough meaningful information to be valuable...